File: app/src/main/java/be/digitalia/fosdem/adapters/TrackScheduleAdapter.java
package be.digitalia.fosdem.adapters;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.LayerDrawable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.text.DateFormat;
import java.util.List;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.core.content.ContextCompat;
import androidx.core.util.ObjectsCompat;
import androidx.core.widget.TextViewCompat;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;
import be.digitalia.fosdem.R;
import be.digitalia.fosdem.model.Event;
import be.digitalia.fosdem.model.StatusEvent;
import be.digitalia.fosdem.utils.DateUtils;

public class TrackScheduleAdapter extends ListAdapter<StatusEvent, TrackScheduleAdapter.ViewHolder> {

	public interface EventClickListener {
		void onEventClick(int position, Event event);
	}

	private static final DiffUtil.ItemCallback<StatusEvent> DIFF_CALLBACK = new SimpleItemCallback<StatusEvent>() {
		@Override
		public boolean areContentsTheSame(@NonNull StatusEvent oldItem, @NonNull StatusEvent newItem) {
			final Event oldEvent = oldItem.getEvent();
			final Event newEvent = newItem.getEvent();
			return ObjectsCompat.equals(oldEvent.getTitle(), newEvent.getTitle())
					&& ObjectsCompat.equals(oldEvent.getPersonsSummary(), newEvent.getPersonsSummary())
					&& ObjectsCompat.equals(oldEvent.getRoomName(), newEvent.getRoomName())
					&& ObjectsCompat.equals(oldEvent.getStartTime(), newEvent.getStartTime())
					&& oldItem.isBookmarked() == newItem.isBookmarked();
		}
	};
	private static final Object TIME_COLORS_PAYLOAD = new Object();
	private static final Object SELECTION_PAYLOAD = new Object();

	final DateFormat timeDateFormat;
	final int timeBackgroundColor;
	final int timeForegroundColor;
	final int timeRunningBackgroundColor;
	final int timeRunningForegroundColor;
	@Nullable
	final EventClickListener listener;

	private long currentTime = -1L;
	private long selectedId = -1L;

	public TrackScheduleAdapter(Context context, @Nullable EventClickListener listener) {
		super(DIFF_CALLBACK);
		setHasStableIds(true);
		timeDateFormat = DateUtils.getTimeDateFormat(context);
		timeBackgroundColor = ContextCompat.getColor(context, R.color.schedule_time_background);
		timeRunningBackgroundColor = ContextCompat.getColor(context, R.color.schedule_time_running_background);

		TypedArray a = context.getTheme().obtainStyledAttributes(R.styleable.PrimaryTextColors);
		timeForegroundColor = a.getColor(R.styleable.PrimaryTextColors_android_textColorPrimary, 0);
		timeRunningForegroundColor = a.getColor(R.styleable.PrimaryTextColors_android_textColorPrimaryInverse, 0);
		a.recycle();

		this.listener = listener;
	}

	/**
	 * @return The position of the item id in the current data set, or -1 if not found.
	 */
	public int getPositionForId(long id) {
		if (id != -1) {
			final int count = getItemCount();
			for (int i = 0; i < count; ++i) {
				if (getItemId(i) == id) {
					return i;
				}
			}
		}
		return RecyclerView.NO_POSITION;
	}

	public void setCurrentTime(long time) {
		if (currentTime != time) {
			currentTime = time;
			notifyItemRangeChanged(0, getItemCount(), TIME_COLORS_PAYLOAD);
		}
	}

	public void setSelectedId(long newId) {
		final long oldId = selectedId;
		if (oldId != newId) {
			selectedId = newId;
			final int count = getItemCount();
			for (int i = 0; i < count; ++i) {
				final long id = getItemId(i);
				if (id == oldId || id == newId) {
					notifyItemChanged(i, SELECTION_PAYLOAD);
				}
			}
		}
	}

	@Override
	public long getItemId(int position) {
		return getItem(position).getEvent().getId();
	}

	@NonNull
	@Override
	public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_schedule_event, parent, false);
		return new ViewHolder(view, R.drawable.activated_background);
	}

	@Override
	public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
		final StatusEvent statusEvent = getItem(position);
		final Event event = statusEvent.getEvent();
		holder.bind(event, statusEvent.isBookmarked());
		holder.bindTimeColors(event, currentTime);
		holder.bindSelection(event.getId() == selectedId);
	}

	@Override
	public void onBindViewHolder(@NonNull ViewHolder holder, int position, @NonNull List<Object> payloads) {
		if (payloads.isEmpty()) {
			onBindViewHolder(holder, position);
		} else {
			final StatusEvent statusEvent = getItem(position);
			if (payloads.contains(TIME_COLORS_PAYLOAD)) {
				holder.bindTimeColors(statusEvent.getEvent(), currentTime);
			}
			if (payloads.contains(SELECTION_PAYLOAD)) {
				holder.bindSelection(statusEvent.getEvent().getId() == selectedId);
			}
		}
	}

	class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		final TextView time;
		final TextView title;
		final TextView persons;
		final TextView room;

		Event event;

		ViewHolder(@NonNull View itemView, @DrawableRes int activatedBackgroundResId) {
			super(itemView);
			time = itemView.findViewById(R.id.time);
			title = itemView.findViewById(R.id.title);
			persons = itemView.findViewById(R.id.persons);
			room = itemView.findViewById(R.id.room);
			itemView.setOnClickListener(this);
			if (activatedBackgroundResId != 0) {
				// Compose a new background drawable by combining the existing one with the activated background
				final Drawable existingBackground = itemView.getBackground();
				final Drawable activatedBackground = ContextCompat.getDrawable(itemView.getContext(), activatedBackgroundResId);
				Drawable newBackground;
				if (existingBackground == null) {
					newBackground = activatedBackground;
				} else {
					// Clear the existing background drawable callback so it can be assigned to the LayerDrawable
					itemView.setBackground(null);
					newBackground = new LayerDrawable(new Drawable[]{existingBackground, activatedBackground});
				}
				itemView.setBackground(newBackground);
			}
		}

		void bind(@NonNull Event event, boolean isBookmarked) {
			Context context = itemView.getContext();
			this.event = event;

			time.setText(timeDateFormat.format(event.getStartTime()));
			title.setText(event.getTitle());
			Drawable bookmarkDrawable = isBookmarked
					? AppCompatResources.getDrawable(context, R.drawable.ic_bookmark_24dp)
					: null;
			TextViewCompat.setCompoundDrawablesRelativeWithIntrinsicBounds(title, null, null, bookmarkDrawable, null);
			title.setContentDescription(isBookmarked
					? context.getString(R.string.in_bookmarks_content_description, event.getTitle())
					: null
			);
			String personsSummary = event.getPersonsSummary();
			persons.setText(personsSummary);
			persons.setVisibility(TextUtils.isEmpty(personsSummary) ? View.GONE : View.VISIBLE);
			room.setText(event.getRoomName());
			room.setContentDescription(context.getString(R.string.room_content_description, event.getRoomName()));
		}

		void bindTimeColors(@NonNull Event event, long currentTime) {
			if ((currentTime != -1L) && event.isRunningAtTime(currentTime)) {
				// Contrast colors for running event
				time.setBackgroundColor(timeRunningBackgroundColor);
				time.setTextColor(timeRunningForegroundColor);
				time.setContentDescription(time.getContext().getString(R.string.in_progress_content_description, time.getText()));
			} else {
				// Normal colors
				time.setBackgroundColor(timeBackgroundColor);
				time.setTextColor(timeForegroundColor);
				// Use text as content description
				time.setContentDescription(null);
			}
		}

		void bindSelection(boolean isSelected) {
			itemView.setActivated(isSelected);
		}

		@Override
		public void onClick(View v) {
			if (listener != null) {
				listener.onEventClick(getAdapterPosition(), event);
			}
		}
	}
}

File: app/src/main/java/be/digitalia/fosdem/adapters/TrackScheduleAdapter.kt
package be.digitalia.fosdem.adapters

import android.content.Context
import android.graphics.drawable.LayerDrawable
import android.view.ContextThemeWrapper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.core.view.isGone
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import be.digitalia.fosdem.R
import be.digitalia.fosdem.model.Event
import be.digitalia.fosdem.model.StatusEvent
import be.digitalia.fosdem.utils.DateUtils
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

class TrackScheduleAdapter(context: Context, private val clickListener: (Event) -> Unit)
    : ListAdapter<StatusEvent, TrackScheduleAdapter.ViewHolder>(EventsAdapter.DIFF_CALLBACK) {

    private val timeFormatter = DateUtils.getTimeFormatter(context)

    var timeZoneOverride: ZoneId? = null
        set(value) {
            if (field != value) {
                field = value
                notifyItemRangeChanged(0, itemCount, TIME_PAYLOAD)
            }
        }

    var currentTime: Instant? = null
        set(value) {
            val oldTime = field
            if (field != value) {
                field = value
                val list = currentList
                val oldIndex = getOngoingIndexForTime(list, oldTime)
                val newIndex = getOngoingIndexForTime(list, value)
                if (oldIndex != newIndex) {
                    if (oldIndex >= 0) notifyItemChanged(oldIndex)
                    if (newIndex >= 0) notifyItemChanged(newIndex)
                }
            }
        }

    private fun getOngoingIndexForTime(list: List<StatusEvent>, time: Instant?): Int {
        if (time == null) {
            return -1
        }
        var index = list.binarySearchBy(time) { it.event.startTime }
        if (index >= 0) {
            return index
        }
        // Check if the event at previous position is in progress
        index = -index - 2
        return if (index >= 0 && list[index].event.isRunningAtTime(time)) index else -1
    }

    var selectedId: Long = RecyclerView.NO_ID
        set(value) {
            val oldId = field
            if (oldId != value) {
                field = value
                currentList.forEachIndexed { index, statusEvent ->
                    val id = statusEvent.event.id
                    if (id == oldId || id == value) {
                        notifyItemChanged(index, SELECTION_PAYLOAD)
                    }
                }
            }
        }

    override fun getItemViewType(position: Int): Int {
        val time = currentTime
        return if (time != null && getItem(position).event.isRunningAtTime(time))
            ONGOING_VIEW_TYPE else DEFAULT_VIEW_TYPE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val context = if (viewType == ONGOING_VIEW_TYPE)
            ContextThemeWrapper(parent.context, R.style.ThemeOverlay_App_OngoingEvent)
        else parent.context
        val view =
            LayoutInflater.from(context).inflate(R.layout.item_schedule_event, parent, false)
        return ViewHolder(view, timeFormatter, R.drawable.activated_background, clickListener)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val statusEvent = getItem(position)
        val event = statusEvent.event
        holder.bind(event, statusEvent.isBookmarked)
        holder.bindTime(event, timeZoneOverride)
        holder.bindSelection(event.id == selectedId)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
        if (payloads.isEmpty()) {
            onBindViewHolder(holder, position)
        } else {
            val statusEvent = getItem(position)
            if (TIME_PAYLOAD in payloads) {
                holder.bindTime(statusEvent.event, timeZoneOverride)
            }
            if (SELECTION_PAYLOAD in payloads) {
                holder.bindSelection(statusEvent.event.id == selectedId)
            }
        }
    }

    class ViewHolder(
        itemView: View, private val timeFormatter: DateTimeFormatter,
        @DrawableRes private val activatedBackgroundResId: Int, private val clickListener: (Event) -> Unit
    ) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private val time: TextView = itemView.findViewById(R.id.time)
        private val title: TextView = itemView.findViewById(R.id.title)
        private val persons: TextView = itemView.findViewById(R.id.persons)
        private val room: TextView = itemView.findViewById(R.id.room)

        private var event: Event? = null

        init {
            itemView.setOnClickListener(this)
            if (activatedBackgroundResId != 0) {
                // Compose a new background drawable by combining the existing one with the activated background
                val existingBackground = itemView.background
                val activatedBackground =
                    ContextCompat.getDrawable(itemView.context, activatedBackgroundResId)
                val newBackground = if (existingBackground == null) {
                    activatedBackground
                } else {
                    // Clear the existing background drawable callback so it can be assigned to the LayerDrawable
                    itemView.background = null
                    LayerDrawable(arrayOf(existingBackground, activatedBackground))
                }
                itemView.background = newBackground
            }
        }

        fun bind(event: Event, isBookmarked: Boolean) {
            val context = itemView.context
            this.event = event

            title.text = event.title
            val bookmarkDrawable = if (isBookmarked) AppCompatResources.getDrawable(context, R.drawable.ic_bookmark_white_24dp) else null
            title.setCompoundDrawablesRelativeWithIntrinsicBounds(null, null, bookmarkDrawable, null)
            title.contentDescription = if (isBookmarked) {
                context.getString(R.string.in_bookmarks_content_description, event.title.orEmpty())
            } else null
            val personsSummary = event.personsSummary
            persons.text = personsSummary
            persons.isGone = personsSummary.isNullOrEmpty()
            room.text = event.roomName
            room.contentDescription = context.getString(R.string.room_content_description, event.roomName.orEmpty())
        }

        fun bindTime(event: Event, timeZoneOverride: ZoneId?) {
            time.text = event.startTime(timeZoneOverride)?.format(timeFormatter) ?: "?"
            if (itemViewType == ONGOING_VIEW_TYPE) {
                time.contentDescription =
                    time.context.getString(R.string.in_progress_content_description, time.text)
            }
        }

        fun bindSelection(isSelected: Boolean) {
            itemView.isActivated = isSelected
        }

        override fun onClick(v: View) {
            event?.let { clickListener(it) }
        }
    }

    companion object {
        private const val DEFAULT_VIEW_TYPE = 0
        private const val ONGOING_VIEW_TYPE = 1
        private val TIME_PAYLOAD = Any()
        private val SELECTION_PAYLOAD = Any()
    }
}